Fixups for #47383 (fixes `runbenchmarks("sort")`)

LilithHafner opened the pull request:
- Test and fix overflow in `CountSort`
- Test and fix overly strict type signature in `setindex!(v::WithoutMissingVector, x, i)`

I believe this is the issue behind https://github.com/JuliaCI/BaseBenchmarks.jl/pull/305, https://github.com/JuliaLang/julia/pull/47795, etc.

Commit:
add test demonstrating overflow in countsort

Commit:
fix overflow in countsort

LilithHafner commented:
The problem is a missing `maybe_unsigned` here

https://github.com/JuliaLang/julia/blob/894f1a5aa1412373954bed51da64d3b32e44e15e/base/sort.jl#L833

That should correspond with the line here:

https://github.com/JuliaLang/julia/blob/894f1a5aa1412373954bed51da64d3b32e44e15e/base/sort.jl#L808

I have a fix queued for once CI fails with the new test.

LilithHafner commented:
The new test failed.

Commit:
convert back to signed without overflow error

LilithHafner commented:
Now the new test passes (also 👍)

LilithHafner commented:
@nanosoldier `runbenchmarks("sort")`

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/47822#issuecomment-1340550509) failed.

LilithHafner commented:
@vtjnash, would you post the log of this failure? I thought this would fix it, but perhaps it's a separate bug or another.

vtjnash commented:
Sure, it seems to be some sort of error with loading the package:
```
      From worker 3:    2022-12-07T03:12:38.160 | starting benchscript.jl (STDOUT/STDERR will be redirected to the result folder)                                                                                                                  
      From worker 3:    ERROR: LoadError: CanonicalIndexError: setindex! not defined for Base.Sort.WithoutMissingVector{Int8, Vector{Union{Missing, Int8}}}                                                                                        
      From worker 3:    Stacktrace:                                                                                                                                                                                                                
      From worker 3:      [1] error_if_canonical_setindex(#unused#::IndexCartesian, A::Base.Sort.WithoutMissingVector{Int8, Vector{Union{Missing, Int8}}}, #unused#::Int64)                                                                        
      From worker 3:        @ Base ./abstractarray.jl:1407                                                                                                                                                                                         
      From worker 3:      [2] setindex!                                                                                                                                                                                                            
      From worker 3:        @ ./abstractarray.jl:1396 [inlined]                                                                                                                                                                                    
      From worker 3:      [3] _sort!(v::Base.Sort.WithoutMissingVector{Int8, Vector{Union{Missing, Int8}}}, #unused#::Base.Sort.CountingSort, o::Base.Order.ForwardOrdering, kw::NamedTuple{(:scratch, :lo, :hi, :mn, :mx), Tuple{Nothing, Int64, I
nt64, Int8, Int8}})                                         
      From worker 3:        @ Base.Sort ./sort.jl:846                                                                    
      From worker 3:      [4] _sort!                        
      From worker 3:        @ ./sort.jl:811 [inlined]                                                                    
      From worker 3:      [5] _sort!(v::Base.Sort.WithoutMissingVector{Int8, Vector{Union{Missing, Int8}}}, a::Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.Sort.CountingSort, Base.Sort.ConsiderRadixSort{Base.Sort.RadixSort, Bas
e.Sort.Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}, o::Base.Order.ForwardOrdering, kw::NamedTuple{(:scratch, :lo, :hi), Tuple{Nothing, Int64, Int64}})
      From worker 3:        @ Base.Sort ./sort.jl:787                                                                    
      From worker 3:      [6] _sort!                        
      From worker 3:        @ ./sort.jl:760 [inlined]                                                                    
      From worker 3:      [7] _sort!(v::Base.Sort.WithoutMissingVector{Int8, Vector{Union{Missing, Int8}}}, a::Base.Sort.Small{40, Base.Sort.InsertionSortAlg, Base.Sort.CheckSorted{Base.Sort.ComputeExtrema{Base.Sort.ConsiderCountingSort{Base.S
ort.CountingSort, Base.Sort.ConsiderRadixSort{Base.Sort.RadixSort, Base.Sort.Small{80, Base.Sort.InsertionSortAlg, PartialQuickSort{Missing, Missing, Base.Sort.InsertionSortAlg}}}}}}}, o::Base.Order.ForwardOrdering, kw::NamedTuple{(:scratch, :
lo, :hi), Tuple{Nothing, Int64, Int64}})                    
      From worker 3:        @ Base.Sort ./sort.jl:692                                                                    
      From worker 3:      [8] _sort!                        
      From worker 3:        @ ./sort.jl:668 [inlined]                                                                    
      From worker 3:      [9] _sort!                        
      From worker 3:        @ ./sort.jl:621 [inlined]                                                                    
      From worker 3:     [10] _sort!                        
      From worker 3:        @ ./sort.jl:692 [inlined]                                                                    
      From worker 3:     [11] _sort!                        
      From worker 3:        @ ./sort.jl:637 [inlined]                                                                    
      From worker 3:     [12] _sort!                        
      From worker 3:        @ ./sort.jl:568 [inlined]                                                                    
      From worker 3:     [13] #sort!#28                     
      From worker 3:        @ ./sort.jl:1359 [inlined]                                                                   
      From worker 3:     [14] sort!                         
      From worker 3:        @ ./sort.jl:1352 [inlined]                                                                   
      From worker 3:     [15] sort(v::Vector{Union{Missing, Int8}}; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})                                                                                                            
      From worker 3:        @ Base.Sort ./sort.jl:1385                                                                   
      From worker 3:     [16] sort                          
      From worker 3:        @ ./sort.jl:1385 [inlined]                                                                   
      From worker 3:     [17] var"##core#9673"(A#9309::Vector{Union{Missing, Int8}})                                     
      From worker 3:        @ BaseBenchmarks.UnionBenchmarks ~/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:489                                                                                                                           
      From worker 3:     [18] var"##sample#9674"(::Tuple{Vector{Union{Missing, Int8}}}, __params::BenchmarkTools.Parameters)                                                                                                                       
      From worker 3:        @ BaseBenchmarks.UnionBenchmarks ~/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:495                                                                                                                           
      From worker 3:     [19] _run(b::BenchmarkTools.Benchmark, p::BenchmarkTools.Parameters; verbose::Bool, pad::String, kwargs::Base.Pairs{Symbol, Integer, NTuple{4, Symbol}, NamedTuple{(:samples, :evals, :gctrial, :gcsample), Tuple{Int64, I
nt64, Bool, Bool}}})                                        
      From worker 3:        @ BenchmarkTools ~/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:99                                                                                                                                            
      From worker 3:     [20] kwcall(::NamedTuple{(:verbose, :pad, :samples, :evals, :gctrial, :gcsample), Tuple{Bool, String, Int64, Int64, Bool, Bool}}, ::typeof(BenchmarkTools._run), b::BenchmarkTools.Benchmark, p::BenchmarkTools.Parameters
)                                                           
      From worker 3:        @ BenchmarkTools ~/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:92                                                                                                                                            
      From worker 3:     [21] #invokelatest#2               
      From worker 3:        @ ./essentials.jl:818 [inlined]                                                              
      From worker 3:     [22] #run_result#45                
      From worker 3:        @ ~/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:34 [inlined]                       
      From worker 3:     [23] run(b::BenchmarkTools.Benchmark, p::BenchmarkTools.Parameters; progressid::Base.UUID, nleaves::Int64, ndone::Int64, kwargs::Base.Pairs{Symbol, Any, NTuple{6, Symbol}, NamedTuple{(:verbose, :pad, :samples, :evals, 
:gctrial, :gcsample), Tuple{Bool, String, Int64, Int64, Bool, Bool}}})                                                   
      From worker 3:        @ BenchmarkTools ~/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:117                                                                                                                                           
      From worker 3:     [24] macro expansion               
      From worker 3:        @ ~/.julia/packages/BenchmarkTools/0owsb/src/execution.jl:135 [inlined]                      
```

Commit:
add another failing test

Commit:
remove unnecessary type annotations (fixes tests)

This fixes the test failure because it allows for automatic conversion.
The manual for implementing the AbstractArray interface also does not recomend
a type signature for the value arg in setindex!.

LilithHafner commented:
The bug is still from #47383; the first pair of commits fixed one bug only to unveil another. The second pair of commits test and fix a second bug. It's wild that these bugs come up in benchmarks but neither CI nor PkgEval (though they will now come up in CI). Perhaps we are benchmarking code paths that are not used anywhere in the package ecosystem.

@nanosoldier `runbenchmarks("sort")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/47822#issuecomment-1342525914) has completed - successfully executed benchmarks. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6a841da/report.md).

petvana reviewed (commented):
I don't see any problem with the PR.
